kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/AdminManager.java

```diff
@@ -439,10 +445,30 @@ private void handleUpdatePartitionsAsync(String topic,
     public void setBrokers(Map<String, Set<Node>> newBrokers) {
         brokersCacheLock.writeLock().lock();
         try {
+            setControllerId(newBrokers);
             this.brokersCache = newBrokers;
         } finally {
             brokersCacheLock.writeLock().unlock();
         }
     }
 
+    // only set when setBrokers
+    private void setControllerId(Map<String, Set<Node>> newBrokers) {
+        Map<String, Integer> newControllerId = Maps.newHashMap();
+        newBrokers.forEach((listenerName, brokers) -> {
+            if (brokers.size() == 0) {
+                newControllerId.put(listenerName, MetadataResponse.NO_CONTROLLER_ID);
+            } else {
+                List<Node> nodes = Lists.newArrayList(brokers);
+                newControllerId.put(listenerName,
+                        nodes.size() > 1 ? nodes.get(random.nextInt(brokers.size())).id() : nodes.get(0).id());
+            }
+        });
+        this.controllerId = newControllerId;
+    }
+
+    // always get the controllerId directly from the cache
+    public int getControllerId(String listenerName) {
```

Comment:
This code is wrong.
This is a HashMap and it is not accessed inside the "lock".
it is strange that spotbugs does not complaint.

btw accessing a HashMap in a non threadsafe way may lead to unpredictable behaviour.
we should fix this
@Demogorgon314 @BewareMyPower

Reply:
There was a discussion in https://github.com/streamnative/kop/pull/1093#discussion_r809048190. Does the explanation of @wenbingshen make sense to you?

From my perspective, the internal state of `controllerId` never changes once it's set with a new value. Since `controllerId` is a volatile field, we can always see the latest object it points to, and the object has only one possible state to watch.

See the only two references of `controllerId`:

```java
    private void setControllerId(Map<String, Set<Node>> newBrokers) {
        /* ... */
        this.controllerId = newControllerId; // [1] points to another map
    }

    public int getControllerId(String listenerName) {
        return controllerId.getOrDefault(listenerName, MetadataResponse.NO_CONTROLLER_ID); // [2] call getOrDefault, which doesn't modify the internal state.
    }
```

Reply:
I believe that it is better to follow the best practices and so do not have bad surprises in production :-)

The "volatile" keyword applies only to the access to the reference to the map, not to the contents. So in a multithread env you are not guaranteed that another thread "sees" the new contents of the map, even if it "sees" the map

we must:
- use ConcurentHashMap e use the "readLock"
- remove "volatile" keyword

I am really sure that not following well established patterns will bring us bad surprises and problems to users

Reply:
The point is, the content could never be published to another thread.

The map is first created here:

```java
    private void setControllerId(Map<String, Set<Node>> newBrokers) {
        Map<String, Integer> newControllerId = Maps.newHashMap();
        /* ... */
        this.controllerId = newControllerId;
    }
```

After it's assigned to `controllerId`, there is no way to modify the content because we only expose the immutable view by `getOrDefault` method.

Reply:
JMM's happens-before rule can guarantee another thread can see the latest content before `[3]` because `[3]` is an assignment of a `volatile` variable so that `[2.1]` and `[2.2]` can be seen by other threads.

```java
   private void setControllerId(Map<String, Set<Node>> newBrokers) {
        Map<String, Integer> newControllerId = Maps.newHashMap(); // [1] create an empty map
        newBrokers.forEach((listenerName, brokers) -> {
            if (brokers.size() == 0) {
                newControllerId.put(listenerName, MetadataResponse.NO_CONTROLLER_ID); // [2.1] put a key-value
            } else {
                List<Node> nodes = Lists.newArrayList(brokers);
                newControllerId.put(listenerName, // [2.2] put a key-value
                        nodes.size() > 1 ? nodes.get(random.nextInt(brokers.size())).id() : nodes.get(0).id());
            }
        });
        this.controllerId = newControllerId; // [3]
```

Reply:
Good. I missed [3].

Thanks for your patience